### PR TITLE
Update Heapster, Monadification, and MRSolver to handle high-order SpecM functions

### PR DESCRIPTION
**Pull request by m-yac**
This PR makes changes to Heapster, Cryptol Monadification, and MRSolver to support higher-order SpecM functions. In particular, this brings things into parity with the update to the Coq definition of the `SpecM` monad which allows such functions to be defined in a sound way (the [`untyped-ho-calls` branch of `entree-specs`](`https://github.com/GaloisInc/entree-specs/tree/untyped-ho-calls`), I believe - @eddywestbrook  is that right?).

Outside of the technical details of Heapster, Monadification, and MRSolver, this PR:
- Moves the Heapster SpecM definitions from `Prelude.sawcore` into their own file, `SpecM.sawcore`, which is also included by default like the Prelude
- Adds an example to `heaspter-saw/examples` which verifies that the top-level interface of the [dilithium reference implementation](https://github.com/pq-crystals/dilithium/tree/standard) satisfies a Cryptol spec using MRSolver

### CODE REVIEW

File: heapster-saw/examples/dilithium.patch
Comment (RyanGlScott) [inline heapster-saw/examples/dilithium.patch:1]: It would be worth including a `README` (or some other form of documentation) which precisely states where the `dilithium` source code is taken from, including a commit hash if relevant. This would be helpful in case we need to reapply the patch in the future.

File: heapster-saw/src/Verifier/SAW/Heapster/LLVMGlobalConst.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/LLVMGlobalConst.hs:41]: Similar comments about this FIXME

File: heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/Permissions.hs:7958]: Minor Haddock fix:

```suggestion
-- \"partial\" because it has not been added to the environment yet
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:404]: Minor Haddock fix:

```suggestion
-- | A \"proof\" that @ctx2@ is an extension of @ctx1@, i.e., that @ctx2@ equals
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:461]: Similar comments about this FIXME

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:475]: Minor Haddock fix:

```suggestion
-- | Use an 'ExprCtxExt' to \"un-extend\" an 'ExprTransCtx'
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:744]: Minor Haddock fix:

```suggestion
-- | Like 'sigmaTypeTransM', but translates @exists x.eq(y)@ into the tuple of
-- types of @x@, omitting the right-hand projection type
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:823]: Similar comments about this FIXME

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:1182]:
```suggestion
-- context of deBruijn indices. Pass the concatenated list of all the kind
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:1475]:
```suggestion
-- A sequence of expressions translates to an ExprTransCtx
```

File: heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/SAWTranslation.hs:6378]: Typo?

```suggestion
-- lambda-bound function indexes in this lambda-abstraction
```

File: heapster-saw/src/Verifier/SAW/Heapster/TypedCrucible.hs
Comment (RyanGlScott) [inline heapster-saw/src/Verifier/SAW/Heapster/TypedCrucible.hs:2515]: Minor Haddock fix:

```suggestion
-- | Helper function to pretty print \"Could not prove ps1 -o ps2\" for
```

File: saw-core/src/Verifier/SAW.hs
Comment (RyanGlScott) [inline saw-core/src/Verifier/SAW.hs:29]: Perhaps this comment should just be removed, as it's a historical note that no longer applies to the current version of the code.

**Review by RyanGlScott — COMMENTED** [review]
Thanks, @m-yac. This is some tremendous work. I lack the context to review this beyond a superficial level, but please let me know if there are individual spots in the code that deserve closer scrutiny.

The two things that are foremost on my mind before landing this are:

* Ensuring that we can figure out the exact revision of the `dilithium` source code we are using (see my review comment below).
* Fixing the CI. It's currently [failing](https://github.com/GaloisInc/saw-script/actions/runs/7634105959/job/20798121412?pr=2017#step:10:19) due to:

  ```
  coq_makefile -f _CoqProject -o Makefile.coq
  (mkdir -p generated/CryptolToCoq; cd ../saw; /home/runner/work/saw-script/saw-script/dist/bin/saw generate_scaffolding.saw)
  (mkdir -p generated/CryptolToCoq; cd ../saw; /home/runner/work/saw-script/saw-script/dist/bin/saw generate_scaffolding.saw)
  [02:01:35.784] Loading file "/home/runner/work/saw-script/saw-script/saw-core-coq/saw/generate_scaffolding.saw"
  [02:01:35.800] Loading file "/home/runner/work/saw-script/saw-script/saw-core-coq/saw/generate_scaffolding.saw"
  COQDEP VFILES
  *** Warning: in file handwritten/CryptolToCoq/SpecM.v, library EnTreeSpecs is required from root EnTree and has not been found in the loadpath!
  *** Warning: in file handwritten/CryptolToCoq/SpecM.v, library TpDesc is required from root EnTree and has not been found in the loadpath!
  *** Warning: in file handwritten/CryptolToCoq/SAWCoreScaffolding.v, library EnTreeSpecs is required from root EnTree and has not been found in the loadpath!
  *** Warning: in file handwritten/CryptolToCoq/SAWCoreVectorsAsCoqVectors.v, library EnTreeSpecs is required from root EnTree and has not been found in the loadpath!
  make[1]: Circular generated/CryptolToCoq/SpecMPrimitivesForSAWCore.vo <- generated/CryptolToCoq/SpecMPrimitivesForSAWCore.vo dependency dropped.
  make[1]: Circular generated/CryptolToCoq/SpecMPrimitivesForSAWCore.vo <- generated/CryptolToCoq/SAWCorePrelude.vo dependency dropped.
  make[1]: Circular handwritten/CryptolToCoq/SAWCoreBitvectors.vo <- generated/CryptolToCoq/SAWCorePrelude.vo dependency dropped.
  make[1]: Circular handwritten/CryptolToCoq/SAWCorePreludeExtra.vo <- generated/CryptolToCoq/SAWCorePrelude.vo dependency dropped.
  make[1]: Circular generated/CryptolToCoq/CryptolPrimitivesForSAWCore.vo <- generated/CryptolToCoq/SpecMPrimitivesForSAWCore.vo dependency dropped.
  make[1]: Circular generated/CryptolToCoq/CryptolPrimitivesForSAWCore.vo <- generated/CryptolToCoq/SAWCorePrelude.vo dependency dropped.
  COQC handwritten/CryptolToCoq/SAWCoreScaffolding.v
  COQC handwritten/CryptolToCoq/CoqVectorsExtra.v
  File "./handwritten/CryptolToCoq/SAWCoreScaffolding.v", line 9, characters 0-39:
  Error: Cannot find a physical path bound to logical path
  EnTreeSpecs with prefix EnTree.
  ```

  If I'm reading [your comment](https://github.com/GaloisInc/saw-script/pull/2017#issue-2097253140) correctly, then this work depends on a `untyped-ho-calls` branch of `entree-specs` that hasn't yet been merged upstream into `main`. Is that right? If so, we should merge that into `main` and update SAW's `entree-specs` dependency [here](https://github.com/GaloisInc/saw-script/blob/adce60e7c5cfb40f9e21918056966914b0965642/.github/workflows/ci.yml#L334-L336) and [here](https://github.com/GaloisInc/saw-script/blob/adce60e7c5cfb40f9e21918056966914b0965642/saw-core-coq/README.md?plain=1#L34).

I realize this is a lot to ask, so please don't hesitate to let me know if there's something I can do to help.

**Comment by m-yac** [comment]
Thank you so much Ryan for taking the time to review this! I don't know the status of that `entree-specs` branch, so I made a PR (https://github.com/GaloisInc/entree-specs/pull/3) to merge that branch into `main` pending @eddywestbrook's approval. In the meantime, I'm going to update the `entree-specs` commit in those places you mentioned to the `HEAD` of that branch in order to make sure the CI is passing, and also review your comments.

File: heapster-saw/examples/dilithium.patch
Comment (m-yac) [reply heapster-saw/examples/dilithium.patch:1]: I added some more info to `dilithium2.saw` - let me know if you think that's enough!

File: cryptol-saw-core/src/Verifier/SAW/Cryptol/Monadify.hs
Comment (m-yac) [reply cryptol-saw-core/src/Verifier/SAW/Cryptol/Monadify.hs:80]: @eddywestbrook What do you think? I like the idea about making issues for these `FIXME`s, but if you think some of them are easy to fix and you point me at them then I'm happy to make those changes. The `FIXME`s that I've personally added are all things that can't be fixed now, so are probably best served as issues.

File: heapster-saw/examples/Dilithium2.saw
Comment (RyanGlScott) [inline heapster-saw/examples/Dilithium2.saw:9]: Should the `LLVM_COMPILER=clang` bit be part of the `Makefile` target for compiling Dilithium?

File: heapster-saw/examples/Makefile
Comment (RyanGlScott) [inline heapster-saw/examples/Makefile:42]: Worth mentioning that if you update the commit hash, you should also update the commit hash in `Dilithium2.saw`.

File: heapster-saw/examples/Dilithium2.saw
Comment (m-yac) [reply heapster-saw/examples/Dilithium2.saw:9]: I'm not sure exactly what you mean by this. Perhaps you're thinking this call to `make` uses the `Makefile` from `heapster-saw/examples`, where the `dilithium` target is? It is not, this is using the `Makefile` of the `dilithium` repo (thus the "in the `ref` directory" bit). Otherwise I'm not sure, because this command [is part of the `Dilithium.bc` target](https://github.com/GaloisInc/saw-script/blob/2ac59d64affacb4620f9ea0108cea5aa15371dd8/heapster-saw/examples/Makefile#L49) in the `heapster-saw/examples` `Makefile`.

Comment (RyanGlScott) [reply heapster-saw/examples/Dilithium2.saw:9]: Silly me, I looked in the `dilithium` target, not the `dilithium%.bc` target (which wasn't visible in the diff). Sorry for the noise!

**Comment by eddywestbrook** [comment]
@RyanGlScott Regarding the entree-specs repo, the only reason I haven't merged that branch into master is because I know the master of SAW is using the current master. But it looks like all of the references to entree-specs in SAW use a specific commit, so it should be safe to just do the merge... If not, I guess things will start failing in the SAW CI!

**Comment by eddywestbrook** [comment]
@RyanGlScott Could you please re-review this? Or are there still things you think I should address?

File: heapster-saw/examples/_CoqProject
Comment (RyanGlScott) [inline heapster-saw/examples/_CoqProject:7]: Do you plan to re-enable `linked_list_proofs.v` (or one of the other tests removed below) at some point? If so, it would be worth tracking this in an issue, along with a description of what remains to be done (i.e., why doesn't it work with the new version of MRSolver?)

File: heapster-saw/examples/clearbufs.saw
Comment (RyanGlScott) [inline heapster-saw/examples/clearbufs.saw:7]: This sort of thing would also be worth mentioning in an issue about re-enabling Heapster tests (should you choose to file one), along with the other FIXMEs in the `heapster-saw/examples` subdirectory.

**Review by RyanGlScott — APPROVED** [review]
Thanks, @eddywestbrook. Two minor comments, but otherwise this LGTM.

File: heapster-saw/examples/_CoqProject
Comment (eddywestbrook) [reply heapster-saw/examples/_CoqProject:7]: Yes, that's a good point, I'll file an issue

**Pull request merged**
